File: examples/dotnet/common/ApplicationPacket.cs
// Copyright (c) 2026 solar
// SPDX-License-Identifier: MIT

using System;

namespace RdplibExample.Common
{
    // The application protocol uses a 2 byte little endian opcode followed by
    // an opcode-specific payload. RDP treats the complete buffer as one message.
    internal sealed class ApplicationPacket
    {
        private const int OpcodeSize = 2;

        internal ApplicationOpcode Opcode { get; private set; }
        internal byte[] Payload { get; private set; }
        internal byte[] Data { get; private set; }

        internal ApplicationPacket(ApplicationOpcode opcode, byte[] payload = null)
        {
            Opcode = opcode;
            Payload = payload ?? Array.Empty<byte>();
            Data = new byte[OpcodeSize + Payload.Length];
            Data[0] = (byte)((ushort)opcode & 0xff);
            Data[1] = (byte)((ushort)opcode >> 8);
            if (Payload.Length != 0)
                Buffer.BlockCopy(Payload, 0, Data, OpcodeSize, Payload.Length);
        }

        private ApplicationPacket()
        {
            Payload = Array.Empty<byte>();
            Data = Array.Empty<byte>();
        }

        internal static bool TryDecode(byte[] data, out ApplicationPacket packet)
        {
            packet = null;
            if (data == null || data.Length < OpcodeSize)
                return false;

            ApplicationPacket decoded = new ApplicationPacket();
            decoded.Opcode = (ApplicationOpcode)(data[0] | (data[1] << 8));
            decoded.Data = data;
            decoded.Payload = new byte[data.Length - OpcodeSize];
            if (decoded.Payload.Length != 0)
                Buffer.BlockCopy(data, OpcodeSize, decoded.Payload, 0, decoded.Payload.Length);

            packet = decoded;
            return true;
        }
    }
}

File: examples/cpp/common/application_packet.cpp
// Copyright (c) 2026 solar
// SPDX-License-Identifier: MIT

#include "application_packet.h"

#include <cstring>

ApplicationPacket::ApplicationPacket()
    : m_opcode(0)
{
}

ApplicationPacket::ApplicationPacket(std::uint16_t opcode, const void *payload, std::uint32_t payload_bytes)
    : m_opcode(opcode),
      m_buffer(OpcodeSize + payload_bytes)
{
    m_buffer[0] = static_cast<std::uint8_t>(opcode & 0xffu);
    m_buffer[1] = static_cast<std::uint8_t>(opcode >> 8u);

    if (payload != nullptr && payload_bytes != 0)
        std::memcpy(m_buffer.data() + OpcodeSize, payload, payload_bytes);
}

bool ApplicationPacket::Decode(const void *data, std::uint32_t bytes)
{
    if (data == nullptr || bytes < OpcodeSize)
        return false;

    const auto *packet = static_cast<const std::uint8_t *>(data);
    m_opcode = static_cast<std::uint16_t>(packet[0]) | static_cast<std::uint16_t>(packet[1] << 8u);
    m_buffer.assign(packet, packet + bytes);
    return true;
}

std::uint16_t ApplicationPacket::Opcode() const
{
    return m_opcode;
}

const std::uint8_t *ApplicationPacket::Payload() const
{
    return m_buffer.size() > OpcodeSize ? m_buffer.data() + OpcodeSize : nullptr;
}

std::uint32_t ApplicationPacket::PayloadSize() const
{
    return m_buffer.size() >= OpcodeSize ? static_cast<std::uint32_t>(m_buffer.size() - OpcodeSize) : 0;
}

const std::uint8_t *ApplicationPacket::Data() const
{
    return m_buffer.empty() ? nullptr : m_buffer.data();
}

std::uint32_t ApplicationPacket::Size() const
{
    return static_cast<std::uint32_t>(m_buffer.size());
}
